glide-core/redis-rs/redis/src/commands/cluster_scan.rs

```diff
@@ -323,51 +367,73 @@ impl ScanState {
         let new_address = if next_slot == END_OF_SCAN {
             return Ok(ScanState::create_finished_state());
         } else {
-            connection.get_address_by_slot(next_slot).await
+            connection
+                .get_address_by_slot(next_slot, &mut scanned_slots_map, allow_non_covered_slots)
+                .await?
         };
-        match new_address {
-            Ok(new_address) => {
-                let new_epoch = connection
-                    .get_address_epoch(&new_address)
-                    .await
-                    .unwrap_or(0);
-                let new_cursor = 0;
-                Ok(ScanState::new(
-                    new_cursor,
-                    scanned_slots_map,
-                    new_address,
-                    new_epoch,
-                    ScanStateStage::InProgress,
-                ))
-            }
-            Err(err) => Err(err),
+
+        if *new_address == NO_ADDRESSES_FOUND {
+            return Ok(ScanState::create_finished_state());
         }
+        let new_epoch = connection
+            .get_address_epoch(&new_address)
+            .await
+            .unwrap_or(0);
+        let new_cursor = 0;
+        Ok(ScanState::new(
+            new_cursor,
+            scanned_slots_map,
+            new_address,
+            new_epoch,
+            ScanStateStage::InProgress,
+        ))
     }
 }
 
 // Implement the [`ClusterInScan`] trait for [`InnerCore`] of async cluster connection.
-#[async_trait]
 impl<C> ClusterInScan for Core<C>
 where
     C: ConnectionLike + Connect + Clone + Send + Sync + 'static,
 {
-    async fn get_address_by_slot(&self, slot: u16) -> RedisResult<Arc<String>> {
-        let address = self
-            .get_address_from_slot(slot, SlotAddr::ReplicaRequired)
-            .await;
-        match address {
-            Some(addr) => Ok(addr),
-            None => {
-                if self.are_all_slots_covered().await {
-                    Err(RedisError::from((
+    async fn get_address_by_slot(
+        &self,
+        mut slot: u16,
+        scanned_slots_map: &mut SlotsBitsArray,
+        allow_non_covered_slots: bool,
+    ) -> RedisResult<Arc<String>> {
+        loop {
+            let address = self
+                .get_address_from_slot(slot, SlotAddr::ReplicaRequired)
+                .await;
+
+            match address {
+                Some(addr) => return Ok(addr),
+                None => {
+                    if allow_non_covered_slots {
```

Comment:
I don't think that this logic should be inside get_address_for_slot, the function should only try to get an address for the given slot, as its name stats, but it actually can increment the slot, return address for other slot, mark slots as done etc.
Please move this logic to the calling function / find a better name that describes what the function does - maybe something like get_next_node

Reply:
get_node_address_for_slot is already exists, wrapping it with another one especially for scan will be redundant, so a better name it is.